// @sinch/numbers/dist/rest/v1/numbers-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumbersService = void 0;
const available_regions_1 = require("./available-regions");
const callbacks_1 = require("./callbacks");
const available_number_1 = require("./available-number");
const active_number_1 = require("./active-number");
/**
 * The Numbers Service exposes the following APIs:
 *  - availableRegions
 *  - availableNumber
 *  - activeNumber
 *  - callbacks
 */
class NumbersService {
    /**
     * Create a new NumbersService instance with its configuration. It needs the following parameters for authentication:
     *  - `projectId`
     *  - `keyId`
     *  - `keySecret`
     *
     * Other supported properties:
     *  - `numbersHostname`
     * @param {SinchClientParameters} params - an Object containing the necessary properties to initialize the service
     */
    constructor(params) {
        this.availableRegions = new available_regions_1.AvailableRegionsApi(params);
        this.callbacks = new callbacks_1.CallbacksApi(params);
        this.availableNumber = new available_number_1.AvailableNumberApi(params);
        this.activeNumber = new active_number_1.ActiveNumberApi(params);
    }
    /**
     * Update the default hostname for each API
     *
     * @param {string} hostname - The new hostname to use for all the APIs.
     */
    setHostname(hostname) {
        this.activeNumber.setHostname(hostname);
        this.availableNumber.setHostname(hostname);
        this.availableRegions.setHostname(hostname);
        this.callbacks.setHostname(hostname);
    }
}
exports.NumbersService = NumbersService;
