// @sinch/sms/dist/rest/v1/batches/batches-api.jest.fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchesApiFixture = void 0;
class BatchesApiFixture {
    constructor() {
        /**
          * Fixture associated to function cancel
          */
        this.cancel = jest.fn();
        /**
          * Fixture associated to function sendDeliveryFeedback
          */
        this.sendDeliveryFeedback = jest.fn();
        /**
          * Fixture associated to function dryRun
          */
        this.dryRun = jest.fn();
        /**
          * Fixture associated to function get
          */
        this.get = jest.fn();
        /**
          * Fixture associated to function list
          */
        this.list = jest.fn();
        /**
          * Fixture associated to function replace
          */
        this.replace = jest.fn();
        /**
          * Fixture associated to function send
          */
        this.send = jest.fn();
        /**
         * Fixture associated to function sendTextMessage
         */
        this.sendTextMessage = jest.fn();
        /**
         * Fixture associated to function send
         */
        this.sendBinaryMessage = jest.fn();
        /**
         * Fixture associated to function send
         */
        this.sendMediaMessage = jest.fn();
        /**
          * Fixture associated to function update
          */
        this.update = jest.fn();
    }
}
exports.BatchesApiFixture = BatchesApiFixture;
