// @sinch/elastic-sip-trunking/dist/rest/v1/elastic-sip-trunking-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticSipTrunkingService = void 0;
const sip_trunks_1 = require("./sip-trunks");
const access_control_list_1 = require("./access-control-list");
const sip_endpoints_1 = require("./sip-endpoints");
const country_permissions_1 = require("./country-permissions");
const calls_history_1 = require("./calls-history");
class ElasticSipTrunkingService {
    constructor(params) {
        this.sipTrunks = new sip_trunks_1.SipTrunksApi(params);
        this.sipEndpoints = new sip_endpoints_1.SipEndpointsApi(params);
        this.accessControlList = new access_control_list_1.AccessControlListApi(params);
        this.countryPermissions = new country_permissions_1.CountryPermissionsApi(params);
        this.calls = new calls_history_1.CallsHistoryApi(params);
    }
    /**
     * Update the default hostname for each API
     *
     * @param {string} hostname - The new hostname to use for all the APIs.
     */
    setHostname(hostname) {
        this.sipTrunks.setHostname(hostname);
        this.sipEndpoints.setHostname(hostname);
        this.accessControlList.setHostname(hostname);
        this.countryPermissions.setHostname(hostname);
        this.calls.setHostname(hostname);
    }
}
exports.ElasticSipTrunkingService = ElasticSipTrunkingService;
