// @sinch/elastic-sip-trunking/dist/rest/v1/sip-trunks/sip-trunks-api.jest.fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SipTrunksApiFixture = void 0;
class SipTrunksApiFixture {
    constructor() {
        /**
         * Fixture associated to function addAccessControlList
         */
        this.addAccessControlList = jest.fn();
        /**
         * Fixture associated to function create
         */
        this.create = jest.fn();
        /**
         * Fixture associated to function deleteAccessControlListFromTrunk
         */
        this.deleteAccessControlList = jest.fn();
        /**
         * Fixture associated to function deleteSipTrunk
         */
        this.delete = jest.fn();
        /**
         * Fixture associated to function getAccessControlListsForTrunk
         */
        this.listAccessControlLists = jest.fn();
        /**
         * Fixture associated to function getSipTrunkById
         */
        this.get = jest.fn();
        /**
         * Fixture associated to function getSipTrunks
         */
        this.list = jest.fn();
        /**
         * Fixture associated to function updateSipTrunk
         */
        this.update = jest.fn();
    }
}
exports.SipTrunksApiFixture = SipTrunksApiFixture;
