// @sinch/sms/dist/rest/v1/sms-domain-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmsDomainApi = void 0;
const sdk_client_1 = require("@sinch/sdk-client");
class SmsDomainApi {
    constructor(sinchClientParameters, apiName) {
        this.sinchClientParameters = sinchClientParameters;
        this.apiName = apiName;
    }
    /**
     * Update the default hostname for the API
     * @param {string} hostname - The new hostname to use for the APIs.
     */
    setHostname(hostname) {
        try {
            this.client = this.getSinchClient();
            this.client.apiClientOptions.hostname = hostname;
        }
        catch (error) {
            console.error('Impossible to set a new hostname, the Application credentials need to be provided first.');
            throw error;
        }
    }
    /**
     * Update the region in the basePath
     * @param {SmsRegion} region - The new region to send the requests to
     */
    setRegion(region) {
        this.sinchClientParameters.smsRegion = region;
        if (this.client) {
            const useZapStack = !this.client.apiClientOptions.useServicePlanId;
            this.client.apiClientOptions.hostname = this.buildHostname(region, useZapStack);
        }
    }
    /**
     * Updates the credentials used to authenticate API requests
     * @param {UnifiedCredentials | ServicePlanIdCredentials} credentials
     */
    setCredentials(credentials) {
        const parametersBackup = { ...this.sinchClientParameters };
        this.sinchClientParameters = {
            ...parametersBackup,
            ...credentials,
        };
        this.resetApiClient();
        try {
            this.getSinchClient();
        }
        catch (error) {
            console.error('Impossible to assign the new credentials to the SMS API');
            this.sinchClientParameters = parametersBackup;
            throw error;
        }
    }
    resetApiClient() {
        this.client = undefined;
    }
    /**
     * Checks the configuration parameters are ok and initialize the API client. Once initialized, the same instance will
     * be returned for the subsequent API calls (singleton pattern)
     * @return {ApiClient} the API Client or throws an error in case the configuration parameters are not ok
     * @private
     */
    getSinchClient() {
        if (!this.client) {
            const region = this.sinchClientParameters.smsRegion ?? sdk_client_1.SmsRegion.UNITED_STATES;
            if (!Object.values(sdk_client_1.SupportedSmsRegion).includes(region)) {
                console.warn(`The region "${region}" is not known as a supported region for the SMS API`);
            }
            const apiClientOptions = (0, sdk_client_1.buildFlexibleOAuth2OrApiTokenApiClientOptions)(this.sinchClientParameters, region, 'SMS');
            this.client = new sdk_client_1.ApiFetchClient(apiClientOptions);
            const useZapStack = !this.client.apiClientOptions.useServicePlanId;
            this.client.apiClientOptions.hostname = this.sinchClientParameters.smsHostname
                ?? this.buildHostname(region, useZapStack);
        }
        return this.client;
    }
    buildHostname(region, useZapStack) {
        const formattedRegion = region !== '' ? `${region}.` : '';
        return (0, sdk_client_1.formatRegionalizedHostname)(sdk_client_1.SMS_HOSTNAME, `${useZapStack ? 'zt.' : ''}${formattedRegion}`);
    }
}
exports.SmsDomainApi = SmsDomainApi;
