// @sinch/sms/dist/rest/v1/sms-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmsService = void 0;
const groups_1 = require("./groups");
const delivery_reports_1 = require("./delivery-reports");
const batches_1 = require("./batches");
const inbounds_1 = require("./inbounds");
/**
 * The SMS Service exposes the following APIs:
 *  - groups
 *  - batches
 *  - inbounds
 *  - deliveryReports
 */
class SmsService {
    /**
     * Create a new SmsService instance with its configuration. This service can be instantiated with 2 different authentication mechanisms:
     * - OAuth2: the required properties are `projectId`, `keyId` and `keySecret`
     * - API Token: the required properties are `servicePlanId` and `apiToken`
     *
     * Other supported properties:
     *  - `smsRegion`
     *  - `smsHostname`
     *  - `forceOAuth2ForSmsApi`
     *  - `forceServicePlanIdUsageForSmsApi`
     * @param {SinchClientParameters} params - an Object containing the necessary properties to initialize the service
     */
    constructor(params) {
        this.groups = new groups_1.GroupsApi(params);
        this.deliveryReports = new delivery_reports_1.DeliveryReportsApi(params);
        this.batches = new batches_1.BatchesApi(params);
        this.inbounds = new inbounds_1.InboundsApi(params);
    }
    /**
     * Update the default hostname for each API
     * @param {string} hostname - The new hostname to use for all the APIs.
     */
    setHostname(hostname) {
        this.groups.setHostname(hostname);
        this.deliveryReports.setHostname(hostname);
        this.batches.setHostname(hostname);
        this.inbounds.setHostname(hostname);
    }
    /**
     * Update the current region for each API
     * @param {SmsRegion} region - The new region to use in the production URL
     */
    setRegion(region) {
        this.groups.setRegion(region);
        this.deliveryReports.setRegion(region);
        this.batches.setRegion(region);
        this.inbounds.setRegion(region);
    }
}
exports.SmsService = SmsService;
