// @sinch/voice/dist/rest/v1/voice-domain-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceDomainApi = void 0;
const sdk_client_1 = require("@sinch/sdk-client");
class VoiceDomainApi {
    constructor(sinchClientParameters, apiName) {
        this.sinchClientParameters = sinchClientParameters;
        this.apiName = apiName;
    }
    /**
     * Update the default hostname for the API
     * @param {string} hostname - The new hostname to use for the APIs.
     */
    setHostname(hostname) {
        try {
            this.client = this.getSinchClient();
            this.client.apiClientOptions.hostname = hostname;
        }
        catch (error) {
            console.error('Impossible to set a new hostname, the Application credentials need to be provided first.');
            throw error;
        }
    }
    /**
     * Update the region in the hostname
     * @param {VoiceRegion} region - The new region to send the requests to
     */
    setRegion(region) {
        this.sinchClientParameters.voiceRegion = region;
        if (this.client) {
            this.client.apiClientOptions.hostname = this.buildHostname(region);
        }
    }
    /**
     * Updates the application credentials used to authenticate API requests
     * @param {ApplicationCredentials} credentials
     */
    setApplication(credentials) {
        const parametersBackup = { ...this.sinchClientParameters };
        this.sinchClientParameters = {
            ...parametersBackup,
            ...credentials,
        };
        this.resetApiClient();
        try {
            this.getSinchClient();
        }
        catch (error) {
            console.error('Impossible to assign the new application to the Voice API');
            this.sinchClientParameters = parametersBackup;
            throw error;
        }
    }
    resetApiClient() {
        this.client = undefined;
    }
    /**
     * Checks the configuration parameters are ok and initialize the API client. Once initialized, the same instance will
     * be returned for the subsequent API calls (singleton pattern)
     * @return {ApiClient} the API Client or throws an error in case the configuration parameters are not ok
     * @private
     */
    getSinchClient() {
        if (!this.client) {
            const apiClientOptions = (0, sdk_client_1.buildApplicationSignedApiClientOptions)(this.sinchClientParameters, 'Voice');
            this.client = new sdk_client_1.ApiFetchClient(apiClientOptions);
            const region = this.sinchClientParameters.voiceRegion ?? sdk_client_1.VoiceRegion.DEFAULT;
            if (!Object.values(sdk_client_1.SupportedVoiceRegion).includes(region)) {
                console.warn(`The region "${region}" is not known as a supported region for the Voice API`);
            }
            this.client.apiClientOptions.hostname = this.buildHostname(region);
        }
        return this.client;
    }
    buildHostname(region) {
        switch (this.apiName) {
            case 'ApplicationsApi':
                return this.sinchClientParameters.voiceApplicationManagementHostname ?? sdk_client_1.VOICE_APPLICATION_MANAGEMENT_HOSTNAME;
            default: {
                const formattedRegion = region === sdk_client_1.VoiceRegion.DEFAULT ? region : `-${region}`;
                return this.sinchClientParameters.voiceHostname ?? (0, sdk_client_1.formatRegionalizedHostname)(sdk_client_1.VOICE_HOSTNAME, formattedRegion);
            }
        }
    }
}
exports.VoiceDomainApi = VoiceDomainApi;
