// @sinch/voice/dist/rest/v1/voice-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceService = void 0;
const applications_1 = require("./applications");
const conferences_1 = require("./conferences");
const calls_1 = require("./calls");
const callouts_1 = require("./callouts");
/**
 * The Voice Service exposes the following APIs:
 * - applications
 * - callouts
 * - conferences
 * - calls
 */
class VoiceService {
    /**
     * Create a new VoiceService instance with its configuration. It needs the following parameters for authentication:
     * - `applicationKey`
     * - `applicationSecret`
     *
     * Other supported properties:
     * - `voiceRegion`
     * - `voiceHostname`
     * - `voiceApplicationManagementHostname`
     * @param {SinchClientParameters} params - an Object containing the necessary properties to initialize the service
     */
    constructor(params) {
        this.applications = new applications_1.ApplicationsApi(params);
        this.conferences = new conferences_1.ConferencesApi(params);
        this.calls = new calls_1.CallsApi(params);
        this.callouts = new callouts_1.CalloutsApi(params);
    }
    /**
     * Update the default hostname for each API except Applications
     * @param {string} hostname - The new hostname to use for all the APIs except Applications.
     */
    setHostname(hostname) {
        this.conferences.setHostname(hostname);
        this.calls.setHostname(hostname);
        this.callouts.setHostname(hostname);
    }
    /**
     * Update the default hostname for the Applications API
     * @param {string} hostname - The new hostname to use for the Applications API.
     */
    setApplicationsManagementHostname(hostname) {
        this.applications.setHostname(hostname);
    }
    /**
     * Update the current region for each API
     * @param {VoiceRegion} region - The new region to use in the production URL
     */
    setRegion(region) {
        this.conferences.setRegion(region);
        this.calls.setRegion(region);
        this.callouts.setRegion(region);
    }
}
exports.VoiceService = VoiceService;
