// @sinch/sdk-client/dist/plugins/exception/exception.response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExceptionResponse = void 0;
const api_errors_1 = require("../../api/api-errors");
/**
 * Plugin to fire an exception on wrong response / data
 */
class ExceptionResponse {
    /**
     * Initialize an instance of the class, with an optional callback function for exception handling.
     *
     * @param {Function} [callback] - A function called in case of an exception. If provided, this function is responsible for throwing the exception or not.
     */
    constructor(callback) {
        this.callback = callback;
    }
    load(context) {
        return {
            transform: (res) => {
                if (context.exception) {
                    return res;
                }
                const errorContext = {
                    apiName: context.apiName,
                    operationId: context.operationId,
                    url: context.url,
                    origin: context.origin,
                };
                let error;
                if (!context.response) {
                    error = new api_errors_1.EmptyResponseError('Fail to Fetch', errorContext, undefined);
                }
                else if (!context.response.ok) {
                    error = new api_errors_1.RequestFailedError(context.response.statusText, context.response.status, errorContext, res);
                }
                else if (!res) {
                    res = {};
                    if (context.response.status !== 204 && context.response.status !== 200) {
                        res = {};
                        error = new api_errors_1.EmptyResponseError(context.response.statusText, errorContext, res);
                    }
                }
                if (error) {
                    if (this.callback) {
                        return this.callback(res, error);
                    }
                    else {
                        throw error;
                    }
                }
                return res;
            },
        };
    }
}
exports.ExceptionResponse = ExceptionResponse;
