// @sinch/fax/dist/rest/v3/fax-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FaxService = void 0;
const emails_1 = require("./emails");
const faxes_1 = require("./faxes");
const services_1 = require("./services");
/**
 * The Fax Service exposes the following APIs:
 *  - services
 *  - faxes
 *  - emails
 */
class FaxService {
    /**
     * Create a new FaxService instance with its configuration. It needs the following parameters for authentication:
     *  - `projectId`
     *  - `keyId`
     *  - `keySecret`
     *
     * Other supported properties:
     *  - `faxRegion`
     *  - `faxHostname`
     * @param {SinchClientParameters} params - an Object containing the necessary properties to initialize the service
     */
    constructor(params) {
        this.emails = new emails_1.EmailsApi(params);
        this.faxes = new faxes_1.FaxesApi(params);
        this.services = new services_1.ServicesApi(params);
    }
    /**
     * Update the default hostname for each API
     * @param {string} hostname - The new hostname to use for all the APIs.
     */
    setHostname(hostname) {
        this.emails.setHostname(hostname);
        this.faxes.setHostname(hostname);
        this.services.setHostname(hostname);
    }
    /**
     * Update the current region for each API
     * @param {FaxRegion} region - The new region to use in the production URL
     */
    setRegion(region) {
        this.emails.setRegion(region);
        this.faxes.setRegion(region);
        this.services.setRegion(region);
    }
}
exports.FaxService = FaxService;
