// @sinch/sdk-client/dist/domain/domain-interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationRegion = exports.SupportedConversationRegion = exports.FaxRegion = exports.SupportedFaxRegion = exports.VoiceRegion = exports.SupportedVoiceRegion = exports.SmsRegion = exports.SupportedSmsRegion = exports.isServicePlanIdCredentials = exports.isUnifiedCredentials = void 0;
const isUnifiedCredentials = (credentials) => {
    const candidate = (credentials);
    return candidate.projectId !== undefined
        && candidate.keyId !== undefined
        && candidate.keySecret !== undefined;
};
exports.isUnifiedCredentials = isUnifiedCredentials;
const isServicePlanIdCredentials = (credentials) => {
    const candidate = (credentials);
    return candidate.servicePlanId !== undefined
        && candidate.apiToken !== undefined;
};
exports.isServicePlanIdCredentials = isServicePlanIdCredentials;
// /////////////
// SMS regions
var SupportedSmsRegion;
(function (SupportedSmsRegion) {
    SupportedSmsRegion["UNITED_STATES"] = "us";
    SupportedSmsRegion["EUROPE"] = "eu";
    SupportedSmsRegion["BRAZIL"] = "br";
    SupportedSmsRegion["CANADA"] = "ca";
    SupportedSmsRegion["AUSTRALIA"] = "au";
})(SupportedSmsRegion || (exports.SupportedSmsRegion = SupportedSmsRegion = {}));
exports.SmsRegion = {
    ...SupportedSmsRegion,
};
// /////////////
// Voice regions
var SupportedVoiceRegion;
(function (SupportedVoiceRegion) {
    SupportedVoiceRegion["DEFAULT"] = "";
    SupportedVoiceRegion["UNITED_STATES"] = "use1";
    SupportedVoiceRegion["EUROPE"] = "euc1";
    SupportedVoiceRegion["SOUTH_AMERICA"] = "sae1";
    SupportedVoiceRegion["SOUTHEAST_ASIA_1"] = "apse1";
    SupportedVoiceRegion["SOUTHEAST_ASIA_2"] = "apse2";
})(SupportedVoiceRegion || (exports.SupportedVoiceRegion = SupportedVoiceRegion = {}));
exports.VoiceRegion = {
    ...SupportedVoiceRegion,
};
// ///////////
// Fax regions
var SupportedFaxRegion;
(function (SupportedFaxRegion) {
    SupportedFaxRegion["DEFAULT"] = "";
    SupportedFaxRegion["UNITED_STATES"] = "use1";
    SupportedFaxRegion["EUROPE"] = "eu1";
    SupportedFaxRegion["SOUTH_AMERICA"] = "sae1";
    SupportedFaxRegion["SOUTHEAST_ASIA_1"] = "apse1";
    SupportedFaxRegion["SOUTHEAST_ASIA_2"] = "apse2";
})(SupportedFaxRegion || (exports.SupportedFaxRegion = SupportedFaxRegion = {}));
exports.FaxRegion = {
    ...SupportedFaxRegion,
};
// ////////////////////
// Conversation regions
var SupportedConversationRegion;
(function (SupportedConversationRegion) {
    SupportedConversationRegion["UNITED_STATES"] = "us";
    SupportedConversationRegion["EUROPE"] = "eu";
    SupportedConversationRegion["BRAZIL"] = "br";
})(SupportedConversationRegion || (exports.SupportedConversationRegion = SupportedConversationRegion = {}));
exports.ConversationRegion = {
    ...SupportedConversationRegion,
};
