// @sinch/sdk-client/dist/domain/domain-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatRegionalizedHostname = exports.VOICE_APPLICATION_MANAGEMENT_HOSTNAME = exports.VOICE_HOSTNAME = exports.VERIFICATION_HOSTNAME = exports.SMS_HOSTNAME = exports.NUMBERS_HOSTNAME = exports.FAX_HOSTNAME = exports.ELASTIC_SIP_TRUNKING_HOSTNAME = exports.CONVERSATION_TEMPLATES_HOSTNAME = exports.CONVERSATION_HOSTNAME = exports.AUTH_HOSTNAME = exports.REGION_PATTERN = void 0;
exports.REGION_PATTERN = '{region}';
exports.AUTH_HOSTNAME = 'https://auth.sinch.com';
exports.CONVERSATION_HOSTNAME = `https://${exports.REGION_PATTERN}conversation.api.sinch.com`;
exports.CONVERSATION_TEMPLATES_HOSTNAME = `https://${exports.REGION_PATTERN}template.api.sinch.com`;
exports.ELASTIC_SIP_TRUNKING_HOSTNAME = 'https://elastic-trunking.api.sinch.com';
exports.FAX_HOSTNAME = `https://${exports.REGION_PATTERN}fax.api.sinch.com`;
exports.NUMBERS_HOSTNAME = 'https://numbers.api.sinch.com';
exports.SMS_HOSTNAME = `https://${exports.REGION_PATTERN}sms.api.sinch.com`;
;
exports.VERIFICATION_HOSTNAME = 'https://verification.api.sinch.com';
exports.VOICE_HOSTNAME = `https://calling${exports.REGION_PATTERN}.api.sinch.com`;
exports.VOICE_APPLICATION_MANAGEMENT_HOSTNAME = 'https://callingapi.sinch.com';
const formatRegionalizedHostname = (hostnamePattern, region) => {
    return hostnamePattern.replace(exports.REGION_PATTERN, region);
};
exports.formatRegionalizedHostname = formatRegionalizedHostname;
