// @sinch/sdk-client/dist/plugins/signing/signing.request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigningRequest = void 0;
const request_plugin_1 = require("../core/request-plugin");
const authorization_helper_1 = require("../../utils/authorization.helper");
class SigningRequest {
    constructor(applicationId, applicationSecret) {
        this.applicationId = applicationId;
        this.applicationSecret = applicationSecret;
    }
    getName() {
        return request_plugin_1.RequestPluginEnum.SIGNING_REQUEST;
    }
    load() {
        return {
            transform: (data) => {
                if (!data.method) {
                    throw new Error('The HTTP method must be defined.');
                }
                if (!data.path) {
                    throw new Error('The URL path must be defined.');
                }
                const authorizationHeader = (0, authorization_helper_1.generateAuthorizationHeader)(data.method.toUpperCase(), data.body, data.headers.get('Content-Type'), `x-timestamp:${data.headers.get('x-timestamp')}`, data.path, this.applicationId, this.applicationSecret);
                data.headers.append('Authorization', authorizationHeader);
                return data;
            },
        };
    }
}
exports.SigningRequest = SigningRequest;
