// @sinch/verification/dist/rest/v1/verification-domain-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationDomainApi = void 0;
const sdk_client_1 = require("@sinch/sdk-client");
class VerificationDomainApi {
    constructor(sinchClientParameters, apiName) {
        this.sinchClientParameters = sinchClientParameters;
        this.apiName = apiName;
    }
    /**
     * Update the default hostname for the API
     * @param {string} hostname - The new hostname to use for the APIs.
     */
    setHostname(hostname) {
        try {
            this.client = this.getSinchClient();
            this.client.apiClientOptions.hostname = hostname;
        }
        catch (error) {
            console.error('Impossible to set a new hostname, the Application credentials need to be provided first.');
            throw error;
        }
    }
    /**
     * Updates the application credentials used to authenticate API requests
     * @param {ApplicationCredentials} credentials
     */
    setApplication(credentials) {
        const parametersBackup = { ...this.sinchClientParameters };
        this.sinchClientParameters = {
            ...parametersBackup,
            ...credentials,
        };
        this.resetApiClient();
        try {
            this.getSinchClient();
        }
        catch (error) {
            console.error('Impossible to assign the new application to the Verification API');
            this.sinchClientParameters = parametersBackup;
            throw error;
        }
    }
    resetApiClient() {
        this.client = undefined;
    }
    /**
     * Checks the configuration parameters are ok and initialize the API client. Once initialized, the same instance will
     * be returned for the subsequent API calls (singleton pattern)
     * @return {ApiClient} the API Client or throws an error in case the configuration parameters are not ok
     * @private
     */
    getSinchClient() {
        if (!this.client) {
            const apiClientOptions = (0, sdk_client_1.buildApplicationSignedApiClientOptions)(this.sinchClientParameters, 'Verification');
            this.client = new sdk_client_1.ApiFetchClient(apiClientOptions);
            this.client.apiClientOptions.hostname = this.sinchClientParameters.verificationHostname ?? sdk_client_1.VERIFICATION_HOSTNAME;
        }
        return this.client;
    }
}
exports.VerificationDomainApi = VerificationDomainApi;
