// @sinch/elastic-sip-trunking/dist/rest/v1/access-control-list/access-control-list-api.jest.fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessControlListApiFixture = void 0;
class AccessControlListApiFixture {
    constructor() {
        /**
         * Fixture associated to function addToTrunk
         */
        this.addToTrunk = jest.fn();
        /**
         * Fixture associated to function listForTrunk
         */
        this.listForTrunk = jest.fn();
        /**
         * Fixture associated to function deleteFromTrunk
         */
        this.deleteFromTrunk = jest.fn();
        /**
         * Fixture associated to function addIpRange
         */
        this.addIpRange = jest.fn();
        /**
         * Fixture associated to function create
         */
        this.create = jest.fn();
        /**
         * Fixture associated to function delete
         */
        this.delete = jest.fn();
        /**
         * Fixture associated to function deleteIpRange
         */
        this.deleteIpRange = jest.fn();
        /**
         * Fixture associated to function list
         */
        this.list = jest.fn();
        /**
         * Fixture associated to function listIpRanges
         */
        this.listIpRanges = jest.fn();
        /**
         * Fixture associated to function update
         */
        this.update = jest.fn();
        /**
         * Fixture associated to function updateIpRange
         */
        this.updateIpRange = jest.fn();
    }
}
exports.AccessControlListApiFixture = AccessControlListApiFixture;
