// @sinch/verification/dist/models/v1/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportVerificationByIdentityHelper = exports.reportVerificationByIdHelper = exports.startVerificationHelper = void 0;
exports.startVerificationHelper = {
    buildSmsRequest: (phoneNumber, reference, smsOptions) => {
        return {
            startVerificationWithSmsRequestBody: {
                identity: {
                    type: 'number',
                    endpoint: phoneNumber,
                },
                reference,
                ...(smsOptions !== undefined) ? {
                    smsOptions: {
                        ...smsOptions,
                    },
                } : {},
            },
        };
    },
    buildCalloutRequest: (phoneNumber, reference, locale) => {
        return {
            startVerificationWithCalloutRequestBody: {
                identity: {
                    type: 'number',
                    endpoint: phoneNumber,
                },
                reference,
                ...(locale !== undefined) ? {
                    calloutOptions: {
                        speech: {
                            locale,
                        },
                    },
                } : {},
            },
        };
    },
    buildFlashCallRequest: (phoneNumber, reference, dialTimeout) => {
        return {
            startVerificationWithFlashCallRequestBody: {
                identity: {
                    type: 'number',
                    endpoint: phoneNumber,
                },
                reference,
                ...(dialTimeout !== undefined) ? {
                    flashCallOptions: {
                        dialTimeout,
                    },
                } : {},
            },
        };
    },
    buildSeamlessRequest: (phoneNumber, reference) => {
        return {
            startSeamlessVerificationRequestBody: {
                identity: {
                    type: 'number',
                    endpoint: phoneNumber,
                },
                reference,
            },
        };
    },
};
exports.reportVerificationByIdHelper = {
    buildSmsRequest: (id, code, cli) => {
        return {
            id,
            reportSmsVerificationByIdRequestBody: {
                sms: {
                    code,
                    cli,
                },
            },
        };
    },
    buildCalloutRequest: (id, code) => {
        return {
            id,
            reportCalloutVerificationByIdRequestBody: {
                callout: {
                    code,
                },
            },
        };
    },
    buildFlashCallRequest: (id, cli) => {
        return {
            id,
            reportFlashCallVerificationByIdRequestBody: {
                flashCall: {
                    cli,
                },
            },
        };
    },
};
exports.reportVerificationByIdentityHelper = {
    buildSmsRequest: (identity, code, cli) => {
        return {
            endpoint: identity,
            reportSmsVerificationByIdentityRequestBody: {
                sms: {
                    code,
                    cli,
                },
            },
        };
    },
    buildCalloutRequest: (identity, code) => {
        return {
            endpoint: identity,
            reportCalloutVerificationByIdentityRequestBody: {
                callout: {
                    code,
                },
            },
        };
    },
    buildFlashCallRequest: (identity, cli) => {
        return {
            endpoint: identity,
            reportFlashCallVerificationByIdentityRequestBody: {
                flashCall: {
                    cli,
                },
            },
        };
    },
};
