// @sinch/verification/dist/rest/v1/verifications/verifications-api.jest.fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationsApiFixture = void 0;
class VerificationsApiFixture {
    constructor() {
        /**
         * Fixture associated to function reportSmsById
         */
        this.reportSmsById = jest.fn();
        /**
         * Fixture associated to function reportFlashCallById
         */
        this.reportFlashCallById = jest.fn();
        /**
         * Fixture associated to function reportCalloutById
         */
        this.reportCalloutById = jest.fn();
        /**
         * Fixture associated to function reportSmsByIdentity
         */
        this.reportSmsByIdentity = jest.fn();
        /**
         * Fixture associated to function reportFlashCallByIdentity
         */
        this.reportFlashCallByIdentity = jest.fn();
        /**
         * Fixture associated to function reportCalloutByIdentity
         */
        this.reportCalloutByIdentity = jest.fn();
        /**
         * Fixture associated to function startSms
         */
        this.startSms = jest.fn();
        /**
         * Fixture associated to function startFlashCall
         */
        this.startFlashCall = jest.fn();
        /**
         * Fixture associated to function startCallout
         */
        this.startCallout = jest.fn();
        /**
         * Fixture associated to function startSeamless
         */
        this.startSeamless = jest.fn();
    }
}
exports.VerificationsApiFixture = VerificationsApiFixture;
