// @sinch/sdk-client/dist/plugins/additional-headers/additional-headers.request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdditionalHeadersRequest = exports.buildHeader = void 0;
const request_plugin_1 = require("../core/request-plugin");
/**
 * Builds a Promise containing the new header in the form of an object {key : value}
 * @param {string} key - the header key
 * @param {string} value - the header value
 */
const buildHeader = async (key, value) => ({ [key]: value });
exports.buildHeader = buildHeader;
class AdditionalHeadersRequest {
    constructor(additionalHeaders) {
        this.additionalHeaders = additionalHeaders;
    }
    getName() {
        return request_plugin_1.RequestPluginEnum.ADDITIONAL_HEADER_REQUEST;
    }
    load() {
        return {
            transform: async (data) => {
                const headers = await this.additionalHeaders.headers;
                if (headers) {
                    Object.keys(headers).forEach((key) => data.headers.set(key, headers[key]));
                }
                return data;
            },
        };
    }
}
exports.AdditionalHeadersRequest = AdditionalHeadersRequest;
